agent/src/util/registration.rs

```diff
@@ -0,0 +1,324 @@
+use akri_discovery_utils::discovery::{
+    v0::{
+        registration_server::{Registration, RegistrationServer},
+        Empty, RegisterRequest,
+    },
+    AGENT_REGISTRATION_SOCKET,
+};
+use akri_shared::uds::unix_stream;
+use futures::TryStreamExt;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Instant;
+use tokio::sync::broadcast;
+use tonic::{transport::Server, Request, Response, Status};
+
+/// Maximum amount of time allowed to pass without being able to connect to a discovery handler
+/// without it being removed from the map of registered Discovery Handlers.
+pub const DISCOVERY_HANDLER_OFFLINE_GRACE_PERIOD_SECS: u64 = 300;
+
+/// Fake endpoint that signals a DiscoveryOperator to use an embedded
+/// discovery handler.
+pub const EMBEDDED_DISCOVERY_HANDLER_ENDPOINT: &str = "embedded";
+
+// Map of RegisterRequests for a specific protocol where key is the endpoint of the Discovery Handler
+// and value is whether or not the discovered devices are local.
+pub type ProtocolDiscoveryHandlerMap = HashMap<String, DiscoveryHandlerDetails>;
+
+/// Map of all registered Discovery Handlers where key is protocol
+/// and value is a map of all Discovery Handlers for that protocol.
+pub type RegisteredDiscoveryHandlerMap = Arc<Mutex<HashMap<String, ProtocolDiscoveryHandlerMap>>>;
+
+/// Describes the connectivity status of a Discovery Handler.
+#[derive(PartialEq, Debug, Clone)]
+pub enum DiscoveryHandlerConnectivityStatus {
+    /// Has a client successfully using it
```

Comment:
[HasClient, Online, Offline] isn't really intuitive to me (could totally just be me) ... maybe something like: [Active|Discovering, Waiting|WaitingForHandlers, Offline]

Reply:
Great point! Fixed in e4e58a212c.